### PR TITLE
feature: 방 만들기 기능 구현 완료

### PR DESCRIPTION
**Pull request by yeonise**
### getRemoteAddr()

- `getRemoteAddr()` 을 통해 사용자 IP를 가져옵니다.

``` yml
server:
  forward-headers-strategy: framework
```

[블로그 글](https://velog.io/@haron/Spring-%EC%B5%9C%EC%B4%88-%EC%9A%94%EC%B2%AD-IP-%EB%8A%94-%EC%96%B4%EB%96%BB%EA%B2%8C-%EA%B0%80%EC%A0%B8%EC%98%A4%EB%8A%94-%EA%B1%B8%EA%B9%8C-6aie0x11)을 참고하여 위 설정을 추가했습니다.

### Session Cookie

세션 쿠키와 관련된 로직은 컨트롤러 단에서 처리하고자 했습니다.

- `@CookieValue` 를 사용하여 세션 쿠키가 존재하는지 확인하고, 존재한다면 이미 다른 방에 참가하고 있다고 판단했습니다.
- 방을 성공적으로 생성한 경우에만 response에 생성했던 쿠키를 담아서 보내도록 했습니다. (방 생성 도중 오류가 발생한 경우 쿠키가 전송되지 않습니다.)

###

### RestDocsSupport : Controller test + REST Docs 문서 생성

- `@Import(PropertiesConfig.class)` 를 추가하여 `application-test.yml` 에서 설정한 **cookieProperties**를 가져오도록 했습니다.
- `andExpect()` 를 적극적으로 사용하여 보다 꼼꼼하게 검증하고자 했습니다.

### IntegrationTestSupport : 통합 환경에서의 테스트

- 중요하다고 생각하는 부분만 테스트 케이스로 남기고자 했습니다.

### Asciidoc 문서 형식 커스터마이징

- 사이드바 목록에서 현재 보고 있는 위치가 표시되도록 했습니다.
- 가독성이 높은 한글 폰트로 변경했습니다.
- 테이블에 Optional 컬럼을 추가했습니다.
- 문서와 문서 사이에 공백을 추가했습니다.

### test-report.yml workflow 변경

- 테스트 수행을 위한 임시 redis 컨테이너를 생성하는 단계를 추가했습니다.
- Actions Secret에 application.yml 파일들을 저장했고 이를 프로젝트에 복사하는 단계를 추가했습니다.

### CODE REVIEW

**Comment by yeonise** [comment]
### Test Coverage Report
|Overall Project|94.7% **`-4.71%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|95.24%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[CustomException.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fexception%2FCustomException.java)|100%|:green_apple:|
|[ErrorType.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fexception%2FErrorType.java)|100%|:green_apple:|
|[GlobalExceptionHandler.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fexception%2FGlobalExceptionHandler.java)|100%|:green_apple:|
|[RandomUtil.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Futil%2FRandomUtil.java)|100%|:green_apple:|
|[ResponseResult.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Futil%2Fapi%2FResponseResult.java)|100%|:green_apple:|
|[SingleRoomParticipationViolationException.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fexception%2Froom%2FSingleRoomParticipationViolationException.java)|100%|:green_apple:|
|[Role.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fuser%2FRole.java)|100%|:green_apple:|
|[RoomController.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2Fpresentation%2FRoomController.java)|100%|:green_apple:|
|[RoomService.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2Fapplication%2FRoomService.java)|100%|:green_apple:|
|[CookieProperties.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fconfig%2Fproperty%2FCookieProperties.java)|100%|:green_apple:|
|[WebConfig.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fconfig%2FWebConfig.java)|100%|:green_apple:|
|[PropertiesConfig.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fconfig%2FPropertiesConfig.java)|100%|:green_apple:|
|[RedisConfig.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fconfig%2FRedisConfig.java)|100%|:green_apple:|
|[RoomCode.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2Fdto%2FRoomCode.java)|100%|:green_apple:|
|[Room.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2FRoom.java)|90.74% **`-9.26%`**|:green_apple:|
|[RoomSettings.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2Fdto%2FRoomSettings.java)|87.5% **`-12.5%`**|:green_apple:|
|[User.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fuser%2FUser.java)|86.84% **`-13.16%`**|:green_apple:|
|[ApiResponse.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Futil%2Fapi%2FApiResponse.java)|86.27% **`-13.73%`**|:green_apple:|
|[AddressArgumentResolver.java](https://github.com/mujik-tigers/you-together/blob/be11bd7558b660c3f0b80232512b8fb4b0f84291/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fresolver%2FAddressArgumentResolver.java)|86.11% **`-13.89%`**|:green_apple:|

File: src/main/java/site/youtogether/resolver/AddressArgumentResolver.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+		return hasAddressAnnotation && isStringType;
+	}
+
+	@Override
+	public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest,
+		WebDataBinderFactory binderFactory) throws Exception {
+		HttpServletRequest request = webRequest.getNativeRequest(HttpServletRequest.class);
+
+		assert request != null;
```
Comment (ghkdgus29) [inline src/main/java/site/youtogether/resolver/AddressArgumentResolver.java:28]: assert 를 사용하면 얻는 효과가 뭔가요?

File: src/main/java/site/youtogether/room/dto/RoomSettings.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+public class RoomSettings {
+
+	@NotBlank
+	@Size(min = 1, max = 30)
+	private String title;
+
+	@Range(min = 2, max = 10)
+	private int capacity;
+
+	@Pattern(regexp = "^[0-9a-zA-Z]{5,10}$")
```
Comment (ghkdgus29) [inline src/main/java/site/youtogether/room/dto/RoomSettings.java:23]: 혹시 입력 형식이 잘못되었을때, 어떤 에러 메시지를 응답하나요?

File: src/main/java/site/youtogether/util/RandomUtil.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	 */
+	public static String generateUserNickname() {
+		List<String> samples = List.of(
+			"MysticTiger", "SilverPhoenix", "ElectricWanderer", "CrimsonDragon", "EmeraldSpecter",
+			"MidnightRider", "VelvetWhisperer", "CosmicStrider", "SolarGoddess", "ArcticShadow",
+			"EnigmaticSphinx", "ScarletSorcerer", "CelestialWatcher", "LunarJester", "SapphireDreamer",
+			"GoldenGlider", "CrimsonFalcon", "EchoingWhisper", "EmberPhoenix", "RadiantRebel"
+		);
+
+		int randomIndex = ThreadLocalRandom.current().nextInt(samples.size());
```
Comment (ghkdgus29) [inline src/main/java/site/youtogether/util/RandomUtil.java:35]: 이 부분, 설명해줄 수 있나요?

File: src/test/java/site/youtogether/room/application/RoomServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+			.title("재밌는 쇼츠 같이 보기")
+			.password(null)
+			.build();
+
+		// when
+		RoomCode roomCode = roomService.create(sessionCode, address, roomSettings);
+
+		// then
+		Room room = roomStorage.findById(roomCode.getRoomCode()).orElseThrow();
+		User user = userStorage.findById(sessionCode).orElseThrow();
```
Comment (ghkdgus29) [inline src/test/java/site/youtogether/room/application/RoomServiceTest.java:54]: ㅋㅋㅋㅋ optional 억지로 가져오고 싶으면 그냥 .get 사용해도 될거같아요

**Comment by ghkdgus29** [comment]
잘봤습니다~!!! 
로직, 테스트 코드, 응답이나 예외처리 다방면에서 고민하신 흔적이 보이네요.

역시 yeonise..

File: src/main/java/site/youtogether/resolver/AddressArgumentResolver.java (hunk 1 above)
Comment (yeonise) [reply src/main/java/site/youtogether/resolver/AddressArgumentResolver.java:28]: 초기 단계에서 객체를 확인을 하여 나중에 더 모호한 NullPointerException으로 실패하는 것을 방지합니다.
위 코드는 IDE에서 제안하는 사항 중 하나입니다.

좀 더 개선된 응답을 제공하고자 한다면, 다음과 같이 작성할 수 있습니다.
```suggestion
		if (request == null) {
		    throw new IllegalArgumentException("HttpServletRequest는 null일 수 없습니다");
		}
```

File: src/test/java/site/youtogether/room/application/RoomServiceTest.java (hunk 4 above)
Comment (yeonise) [reply src/test/java/site/youtogether/room/application/RoomServiceTest.java:54]: ㅋㅋㅋㅋ 굿

File: src/main/java/site/youtogether/room/dto/RoomSettings.java (hunk 2 above)
Comment (yeonise) [reply src/main/java/site/youtogether/room/dto/RoomSettings.java:23]: 기본 설정값으로 응답합니다.
더 자세한 응답 메시지를 원하신다면 `message=" "` 를 추가해도 좋습니다.

File: src/main/java/site/youtogether/util/RandomUtil.java (hunk 3 above)
Comment (yeonise) [reply src/main/java/site/youtogether/util/RandomUtil.java:35]: ThreadLocalRandom은 난수를 생성하는 클래스입니다.
각 스레드마다 고유한 난수 생성기를 사용하기 때문에 서로 간섭받지 않고 독립적인 수를 생성할 수 있습니다.
또한 동기화를 사용하지 않기 때문에 Random 클래스보다 성능적인 이점이 있습니다.

- `.current()` 는 현재 스레드에서 사용할 ThreadLocalRandom 인스턴스를 반환합니다.
- `.nextInt(int bound)` 는 bound 내의 난수를 생성합니다. (이때, bound 값을 포함하지 않습니다.)

[참고](https://www.baeldung.com/java-thread-local-random#:~:text=The%20random%20number%20obtained%20by,support%20setting%20the%20seed%20explicitly)

File: src/main/java/site/youtogether/room/dto/RoomSettings.java (hunk 2 above)
Comment (yeonise) [reply src/main/java/site/youtogether/room/dto/RoomSettings.java:23]: 현재 응답 예시입니다.

``` json
{
    "code": 400,
    "status": "Bad Request",
    "result": "예외가 발생했습니다",
    "data": [
        {
            "type": "password",
            "message": "must match \"^[0-9a-zA-Z]{5,10}$\""
        },
        {
            "type": "title",
            "message": "must not be blank"
        },
        {
            "type": "capacity",
            "message": "must be between 2 and 10"
        }
    ]
}
```

Comment (ghkdgus29) [reply src/main/java/site/youtogether/room/dto/RoomSettings.java:23]: 메세지를 추가하는게 더 좋아보입니다~

Comment (yeonise) [reply src/main/java/site/youtogether/room/dto/RoomSettings.java:23]: 다음과 같이 메시지를 추가했습니다.

``` json
{
    "code": 400,
    "status": "Bad Request",
    "result": "예외가 발생했습니다",
    "data": [
        {
            "type": "password",
            "message": "비밀번호는 5자 이상 10자 이하의 영문 또는 숫자로 입력해 주세요"
        },
        {
            "type": "title",
            "message": "공백이 아닌 문자를 1개 이상 입력해 주세요"
        },
        {
            "type": "capacity",
            "message": "정원은 2명 이상 10명 이하로 입력해 주세요"
        }
    ]
}
```

``` json
{
    "code": 400,
    "status": "Bad Request",
    "result": "예외가 발생했습니다",
    "data": [
        {
            "type": "password",
            "message": "비밀번호는 5자 이상 10자 이하의 영문 또는 숫자로 입력해 주세요"
        },
        {
            "type": "title",
            "message": "제목은 1자 이상 30자 이하로 입력해 주세요"
        },
        {
            "type": "capacity",
            "message": "정원은 2명 이상 10명 이하로 입력해 주세요"
        }
    ]
}
```

**Comment by yeonise** [comment]
### Test Coverage Report
|Overall Project|94.7% **`-4.71%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|95.24%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[CustomException.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fexception%2FCustomException.java)|100%|:green_apple:|
|[ErrorType.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fexception%2FErrorType.java)|100%|:green_apple:|
|[GlobalExceptionHandler.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fexception%2FGlobalExceptionHandler.java)|100%|:green_apple:|
|[RandomUtil.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Futil%2FRandomUtil.java)|100%|:green_apple:|
|[ResponseResult.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Futil%2Fapi%2FResponseResult.java)|100%|:green_apple:|
|[SingleRoomParticipationViolationException.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fexception%2Froom%2FSingleRoomParticipationViolationException.java)|100%|:green_apple:|
|[Role.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fuser%2FRole.java)|100%|:green_apple:|
|[RoomController.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2Fpresentation%2FRoomController.java)|100%|:green_apple:|
|[RoomService.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2Fapplication%2FRoomService.java)|100%|:green_apple:|
|[CookieProperties.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fconfig%2Fproperty%2FCookieProperties.java)|100%|:green_apple:|
|[WebConfig.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fconfig%2FWebConfig.java)|100%|:green_apple:|
|[PropertiesConfig.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fconfig%2FPropertiesConfig.java)|100%|:green_apple:|
|[RedisConfig.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fconfig%2FRedisConfig.java)|100%|:green_apple:|
|[RoomCode.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2Fdto%2FRoomCode.java)|100%|:green_apple:|
|[Room.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2FRoom.java)|90.74% **`-9.26%`**|:green_apple:|
|[RoomSettings.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Froom%2Fdto%2FRoomSettings.java)|87.5% **`-12.5%`**|:green_apple:|
|[User.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fuser%2FUser.java)|86.84% **`-13.16%`**|:green_apple:|
|[ApiResponse.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Futil%2Fapi%2FApiResponse.java)|86.27% **`-13.73%`**|:green_apple:|
|[AddressArgumentResolver.java](https://github.com/mujik-tigers/you-together/blob/a8b230dc2bf9d09c7d6379e4aea3fee1e7709f6e/src%2Fmain%2Fjava%2Fsite%2Fyoutogether%2Fresolver%2FAddressArgumentResolver.java)|86.11% **`-13.89%`**|:green_apple:|

**Pull request merged**